lib/kafka_ex/server_kayrock.ex

```diff
@@ -0,0 +1,946 @@
+      coordinator_key: consumer_group,
+      coordinator_type: 0
+    }
+
+    {response, updated_state} = kayrock_network_request(request, :any, state)
+
+    case response do
+      {:ok,
+       %Kayrock.FindCoordinator.V1.Response{
+         error_code: 0,
+         coordinator: coordinator
+       }} ->
+        State.put_consumer_group_coordinator(
+          updated_state,
+          consumer_group,
+          coordinator.node_id
+        )
+
+      error ->
+        Logger.warn(
+          "Unable to find consumer group coordinator for " <>
+            "#{inspect(consumer_group)}: Error " <>
+            "#{Kayrock.ErrorCode.code_to_atom(error)}"
+        )
+
+        updated_state
+    end
+  end
+
+  defp first_broker_response(request, brokers, timeout) do
+    Enum.find_value(brokers, fn broker ->
+      if Broker.connected?(broker) do
+        try_broker(broker, request, timeout)
+      end
+    end)
+  end
+
+  defp try_broker(broker, request, timeout) do
+    case NetworkClient.send_sync_request(broker, request, timeout) do
+      {:error, error} ->
+        Logger.warn("Network call resulted in error: #{inspect(error)}")
+        nil
+
+      response ->
+        response
+    end
+  end
+
+  defp timeout_val(nil) do
+    Application.get_env(:kafka_ex, :sync_timeout, @default_call_timeout)
+  end
+
+  defp timeout_val(timeout) when is_integer(timeout), do: timeout
+
+  defp config_sync_timeout(timeout \\ nil) do
+    timeout || Application.get_env(:kafka_ex, :sync_timeout, @sync_timeout)
+  end
+
+  defp default_partitioner do
+    Application.get_env(:kafka_ex, :partitioner, KafkaEx.DefaultPartitioner)
+  end
+
+  defp consumer_group_if_auto_commit?(true, state), do: consumer_group?(state)
+  defp consumer_group_if_auto_commit?(false, _state), do: true
+
+  # note within the genserver state, we've already validated the
+  # consumer group, so it can only be either :no_consumer_group or a
+  # valid binary consumer group name
+  defp consumer_group?(%State{
+         consumer_group_for_auto_commit: :no_consumer_group
+       }) do
+    false
+  end
+
+  defp consumer_group?(_), do: true
+
+  defp get_api_versions(state, request_version \\ 0) do
+    request = Kayrock.ApiVersions.get_request_struct(request_version)
+
+    {{ok_or_error, response}, state_out} =
+      kayrock_network_request(request, :any, state)
+
+    {ok_or_error, response, state_out}
+  end
+
+  defp kayrock_network_request(
+         request,
+         node_selector,
+         state,
+         network_timeout \\ nil
+       ) do
+    # produce request have an acks field and if this is 0 then we do not want to
+    # wait for a response from the broker
+    synchronous =
+      case Map.get(request, :acks) do
+        0 -> false
+        _ -> true
+      end
+
+    network_timeout = config_sync_timeout(network_timeout)
+
+    {sender, updated_state} =
+      get_sender(node_selector, state, network_timeout, synchronous)
+
+    case sender do
```

Comment:
I would have expected `sender` to be a verb, since it's a function -- it was a bit confusing

Reply:
hm good point, I'll try to think of something better.

Reply:
Renamed to `get_send_request_function`/`send_request`